### DISCUSSION TITLE
Webpack module federation is not working with eager shared libs

### DISCUSSION THREAD

**Discussion by vovkvlad** [webpack/webpack#12747] [Q&A]
I was looking into Webpack 5 Module federation feature, and have some trouble understanding why my code does not work. The idea is pretty similar to what standard module federation examples do:

`app1` - is the host app `app2` - is a remote exposing the whole app to `app1`

(`app1` renders the header and horizontal line, below which the `app2` should be rendered)

Both `app1` and `app2` declares `react` and `react-dom` as their shared, singleton, eager dependencies in the `weback.config.js`:

```js
// app1 webpack.config.js
const packageJsonDeps = require('./package.json').dependencies;

module.exports = {
  entry: path.resolve(SRC_DIR, './index.js');,
  ...
  plugins: [
    new ModuleFederationPlugin({
      name: "app1",
      remotes: {
        app2: `app2@//localhost:2002/remoteEntry.js`,
      },
      shared: {
        ...packageJsonDeps,
        react: { singleton: true, eager: true, requiredVersion: packageJsonDeps.react },
        "react-dom": { singleton: true, eager: true, requiredVersion: packageJsonDeps["react-dom"] }
      },
    }),
    ...
  ],
};
```

```js
// app2 webpack.config.js
const packageJsonDeps = require('./package.json').dependencies;

module.exports = {
  entry: path.resolve(SRC_DIR, './index.js');,
  ...
  plugins: [
    new ModuleFederationPlugin({
      name: "app2",
      library: { type: "var", name: "app2" },
      filename: "remoteEntry.js",
      exposes: {
        "./App": "./src/App",
      },
      shared: {
        ...packageJsonDeps,
        react: { singleton: true, eager: true, requiredVersion: packageJsonDeps.react, },
        "react-dom": { singleton: true, eager: true, requiredVersion: packageJsonDeps["react-dom"] }
      },
    }),
    ...
  ],
};
```

In the App1 index.js I have next code:

```js
import React from "react";
import ReactDOM from "react-dom";
import App from "./App";


ReactDOM.render(<App />, document.getElementById("root"));
```

The App1 `App.js` component is next:

```js
import React, { Suspense } from 'react';

const RemoteApp2 = React.lazy(() => import("app2/App"));

export default function App() {
  return (
    <div>
      <h1>App 1</h1>
      <p>Below will be some content</p>
      <hr/>
      <Suspense fallback={'Loading App 2'}>
        <RemoteApp2 />
      </Suspense>
    </div>
  );
}
```

But when I start the application I get the next error:
```
Uncaught Error: Shared module is not available for eager consumption: webpack/sharing/consume/default/react/react?1bb3
    at Object.__webpack_modules__.<computed> (consumes:133)
    at __webpack_require__ (bootstrap:21)
    at fn (hot module replacement:61)
    at Module../src/index.js (main.bundle.a8d89941f5dd9a37d429.js:239)
    at __webpack_require__ (bootstrap:21)
    at startup:4
    at startup:6
```

If I extract everything from `index.js` to `bootstrap.js` and in `index.js` will do

```js
import('./bootstrap');
```

Everything works just fine.

This confuses me as [official docs][1] and [blog posts][2] from the creator states that you can do either `bootstrap.js` way OR declare dependency as an eager one.

Would appreciate any help/insights on why it does not work without `bootstrap.js` pattern.

Here is a link to full GitHub sandbox I was building: https://github.com/vovkvlad/webpack-module-fedaration-sandbox/tree/master/simple

  [1]: https://webpack.js.org/concepts/module-federation/#troubleshooting
  [2]: https://medium.com/dev-genius/module-federation-advanced-api-inwebpack-5-0-0-beta-17-71cd4d42e534

**Comment by vovkvlad** [comment]
For anyone interested, I asked the same question on StackOverflow. And there are some answers (also they are mainly guesses, but still, can be helpful enough)
https://stackoverflow.com/questions/66123283/webpack-module-federation-is-not-working-with-eager-shared-libs

**Reply by krutoo** [reply]
It still looks like a bug, need to add bootstrap file with `eager` option of shared modules

**Comment by bchenSyd** [comment]
this seems to be a bug to me.
in `app1` , i.e. the host, we have both `react` and `react-dome` as `eager:true` which means they won't be put into a async chunk, but rather will be part of `main` chunk. There fore, there should be NO reason why they are't ready in the initial load;

actually, if you debug the `main.js` file, you would find that both `react` and `react-dom` are loaded (at least it appears to be this way),, it's just some weird logic around `webpack/shared/consumer/xxx` that blows up, which doesn't make sense to me at all

**Comment by bchenSyd** [comment]
I'm confused with the `eager` flag, let's say in remote config
```js
    filename: "remoteEntry.js",
    exposes:{ xxx.  }
    shared: {
        react: {
          eager: true,
          requiredVersion: "^18.2.0",
        },
```

here `react` is bundled into `remoteEntry.js` file and will be loaded synchronously when exposed modules are loaded;

then **how would a host override react??** 

what's the purpose of the eager option?  I know how it behaves, but I don't know the use case of it

**Reply by bchenSyd** [reply]
very murky area with very little documentation... the office doc doesn't provide enough info, , everything is a guess work

**Comment by royling** [comment]
The author of Webpack Module Federation had explained it very clearly in [this video](https://www.youtube.com/watch?v=d1SS7KAsbdY&t=356s) regarding why the bootstrap file.
